File: src/bin/0235-lowest-common-ancestor-of-a-binary-search-tree.py
# @leetup=info id=235 lang=python3 slug=lowest-common-ancestor-of-a-binary-search-tree



from typing import Optional

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


# @leetup=code
class Solution:
    def lowestCommonAncestor(
        self, root: TreeNode, p: TreeNode, q: TreeNode
    ) -> TreeNode:
        if root.val == p.val or root.val == q.val:
            return root
        if min(p.val, q.val) < root.val < max(p.val, q.val):
            return root
        if p.val < root.val:
            return self.lowestCommonAncestor(root.left, p, q)
        return self.lowestCommonAncestor(root.right, p, q)


# @leetup=code

File: src/bin/0235-lowest-common-ancestor-of-a-binary-search-tree.rs
// @leetup=info id=235 lang=rust slug=lowest-common-ancestor-of-a-binary-search-tree


// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

// @leetup=code
use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    pub fn lowest_common_ancestor(
        root: Option<Rc<RefCell<TreeNode>>>,
        p: Option<Rc<RefCell<TreeNode>>>,
        q: Option<Rc<RefCell<TreeNode>>>,
    ) -> Option<Rc<RefCell<TreeNode>>> {
        fn lowest(root: Rc<RefCell<TreeNode>>, p: i32, q: i32) -> Rc<RefCell<TreeNode>> {
            let root_val = root.borrow().val;
            if root_val == p || root_val == q {
                root
            } else if p < root_val && root_val < q {
                root
            } else if p < root_val {
                lowest(root.borrow().left.as_ref().unwrap().clone(), p, q)
            } else {
                lowest(root.borrow().right.as_ref().unwrap().clone(), p, q)
            }
        }

        let p = p.unwrap().borrow().val;
        let q = q.unwrap().borrow().val;
        Some(lowest(root.unwrap(), p.min(q), p.max(q)))
    }
}
// @leetup=code

struct Solution;

fn main() {
    let _ = Solution::lowest_common_ancestor;
}
